src/Features/LanguageServer/ProtocolUnitTests/Ordering/RequestOrderingTests.cs

```diff
@@ -0,0 +1,126 @@
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.UnitTests.Workspaces;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.UnitTests.RequestOrdering
+{
+    public partial class RequestOrderingTests : AbstractLanguageServerProtocolTests
+    {
+        protected override TestComposition Composition => base.Composition
+            .AddParts(typeof(SerialHandler))
+            .AddParts(typeof(ParallelHandler));
+
+        [Fact]
+        public async Task SerialRequestsDontOverlap()
+        {
+            var requests = new[] {
+                new OrderedLspRequest(SerialHandler.MethodName),
+                new OrderedLspRequest(SerialHandler.MethodName),
+                new OrderedLspRequest(SerialHandler.MethodName),
+            };
+
+            var responses = await TestAsync(requests);
+
+            // Every request should have started at or after the one before it
+            Assert.True(responses[1].StartTime >= responses[0].EndTime);
+            Assert.True(responses[2].StartTime >= responses[1].EndTime);
+        }
+
+        [Fact]
+        public async Task ParallelRequestsOverlap()
+        {
+            var requests = new[] {
+                new OrderedLspRequest(ParallelHandler.MethodName),
+                new OrderedLspRequest(ParallelHandler.MethodName),
+                new OrderedLspRequest(ParallelHandler.MethodName),
+            };
+
+            var responses = await TestAsync(requests);
+
+            // Every request should have started immediately, without waiting
+            Assert.True(responses[1].StartTime < responses[0].EndTime);
+            Assert.True(responses[2].StartTime < responses[1].EndTime);
+        }
+
+        [Fact]
+        public async Task ParallelWaitsForSerial()
+        {
+            var requests = new[] {
+                new OrderedLspRequest(SerialHandler.MethodName),
+                new OrderedLspRequest(ParallelHandler.MethodName),
+                new OrderedLspRequest(ParallelHandler.MethodName),
+            };
+
+            var responses = await TestAsync(requests);
+
+            // The parallel tasks should have waited for the first task to finish
+            Assert.True(responses[1].StartTime >= responses[0].EndTime);
+            Assert.True(responses[2].StartTime >= responses[0].EndTime);
+            // The parallel requests shouldn't have waited for each other
+            Assert.True(responses[1].StartTime < responses[2].EndTime);
+            Assert.True(responses[2].StartTime < responses[1].EndTime);
+        }
+
+        [Fact]
+        public async Task SerialWaitsForParallel()
```

Comment:
likely should have a test that produces a *lot* of tasks randomly based on a seed, then checks invariants about it.